### PR TITLE
feat: add <webview>.sendToFrame() / frameId to 'ipc-message' event

### LINKED ISSUE #30207: [Feature Request]: allow webview.send to send to iframes enabled via nodeintegrationinsubframes
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Currently, there is no way to use the Webview.send method to a preloaded script running inside of an iframe of a webview.

As it stands now, webview.send can only send messages to the webview directly. None of the iframes located inside of the webview can be communicated to via webview.send, however they ARE able to communicate to the rendererprocess via their own respective ipcRenderer.sendToHost methods.

### Proposed Solution

Allow webview.send to specify which process inside the webview it wants to send the message to, defaulting to the baseline webview source.

### Alternatives Considered

I believe this may be possible to do with webcontents.sendToFrame, while passing in the frameid, located inside of Webframe.routingId however this is extremely clunky and requires you to enable the remote module for access inside of the renderer process which is not desired

### Additional Information

Preload scripts can only be ran inside iframes located inside webviews via the nodeintegrationinsubframes tag.

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
This allows replying to an IPC sent by a frame (when `nodeIntegrationInSubFrames` is enabled) inside of the `<webview>`:
```js
webview.addEventListener('ipc-message', event => {
  webview.sendToFrame(event.frameId, event.channel, ...event.args);
});
```
Fixes #30207.

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added `<webview>.sendToFrame()` / `frameId` to `'ipc-message'` event.

### CODE REVIEW

**Comment by ryankashi** [comment]
webview.sendToFrame with passing in the frameId looks good to me!

**Review by zcbenz — APPROVED** [review]
API looks good to me.

File: docs/api/webview-tag.md [hunk 1]
```diff
@@ -606,6 +606,23 @@ listening to the `channel` event with the [`ipcRenderer`](ipc-renderer.md) modul
 See [webContents.send](web-contents.md#contentssendchannel-args) for
 examples.
 
+### `<webview>.sendToFrame(frameId, channel, ...args)`
+
+* `frameId` Integer | [number, number] - the ID of the frame to send to, or a
+  pair of `[processId, frameId]` if the frame is in a different process to the
+  main frame.
```
Comment (samuelmaddock) [inline docs/api/webview-tag.md:613]: I'm a bit concerned about setting a precedence for accepting inflexible arguments here which would be difficult to change in the future. I'd feel much more comfortable if we could pass a `WebFrameMain` instance instead.

Part of my concern is that `FrameTreeNode` IDs might be a more stable singular identifier for frames which I might propose using in the future.

When #29290 is eventually merged, we could add a lazy `frame` property directly to the Event object. 
https://github.com/electron/electron/blob/44491b023ac2653538b7ac36bb73f3085a938666/shell/common/gin_helper/event_emitter.cc#L70-L74

Comment (samuelmaddock) [reply docs/api/webview-tag.md:613]: I also wonder whether adding such a convenience method is necessary when [`frame.send`](https://www.electronjs.org/docs/api/web-frame-main#framesendchannel-args) exists already.

With the lazy `frame` property, we could potentially have

```ts
webview.addEventListener('ipc-message', event => {
  event.frame.send(event.channel, ...event.args);
});
```

File: lib/browser/guest-view-manager.ts [hunk 2]
```diff
@@ -155,8 +155,12 @@ const createGuest = function (embedder: Electron.WebContents, embedderFrameId: n
   });
 
   // Dispatch guest's IPC messages to embedder.
-  guest.on('ipc-message-host' as any, function (_: Electron.Event, channel: string, args: any[]) {
-    sendToEmbedder(IPC_MESSAGES.GUEST_VIEW_INTERNAL_IPC_MESSAGE, channel, ...args);
+  guest.on('ipc-message-host' as any, function (event: Electron.IpcMainEvent, channel: string, args: any[]) {
+    sendToEmbedder(IPC_MESSAGES.GUEST_VIEW_INTERNAL_IPC_MESSAGE, {
+      frameId: [event.processId, event.frameId],
```
Comment (samuelmaddock) [inline lib/browser/guest-view-manager.ts:160]: I think I'd be happier  with something like this so we won't need to introduce another API method.
```suggestion
      get frame() {
        return webFrameMain.fromId(event.processId, event.frameId)
      },
```

Comment (samuelmaddock) [reply lib/browser/guest-view-manager.ts:160]: Ah, I understand now that this API is being introduced to the renderer. :)

File: docs/api/webview-tag.md (hunk 1 above)
Comment (samuelmaddock) [reply docs/api/webview-tag.md:613]: With the understanding now that this API is added to the renderer, I think the API looks good overall.

**Review by samuelmaddock — APPROVED** [review]
API LGTM

File: docs/api/webview-tag.md (hunk 1 above)
Comment (miniak) [reply docs/api/webview-tag.md:613]: we cannot use `WebFrameMain` here as it's the renderer process

File: lib/browser/guest-view-manager.ts (hunk 2 above)
Comment (miniak) [reply lib/browser/guest-view-manager.ts:160]: we are emitting the `<webview>` event in the renderer process, there is no `webFrameMain` there

**Review by ckerr — APPROVED** [review]

File: docs/api/webview-tag.md [hunk 3]
```diff
@@ -606,6 +606,23 @@ listening to the `channel` event with the [`ipcRenderer`](ipc-renderer.md) modul
 See [webContents.send](web-contents.md#contentssendchannel-args) for
 examples.
 
+### `<webview>.sendToFrame(frameId, channel, ...args)`
+
+* `frameId` Integer | [number, number] - the ID of the frame to send to, or a
```
Comment (nornagon) [inline docs/api/webview-tag.md:611]: I don't love that we're propagating the mistake here that there exists a globally unique frame id that is a single integer. the `| [number, number]` was added to existing APIs to support backwards compatibility and not break apps, but maybe new APIs should only accept the `[number, number]` version and not accept a single number as a frame ID.

Thoughts?

File: docs/api/webview-tag.md [hunk 4]
```diff
@@ -877,6 +894,7 @@ webview.addEventListener('close', () => {
 
 Returns:
 
+* `frameId` [number, number] - pair of `[processId, frameId]`.
 * `channel` String
 * `args` any[]
```
Comment (nornagon) [inline docs/api/webview-tag.md:899]: Since this is &lt;webview>, this isn't actually a new positional argument, right? It's a new property on the `event` object?

File: shell/renderer/api/electron_api_web_frame.cc [hunk 5]
```diff
@@ -397,7 +397,7 @@ class WebFrameRenderer : public gin::Wrappable<WebFrameRenderer>,
 
  private:
   bool MaybeGetRenderFrame(v8::Isolate* isolate,
-                           const std::string& method_name,
+                           const char* method_name,
```
Comment (nornagon) [inline shell/renderer/api/electron_api_web_frame.cc:400]: This change seems unrelated, would you mind reverting it?

**Review by nornagon — CHANGES_REQUESTED** [review]

File: docs/api/webview-tag.md (hunk 3 above)
Comment (miniak) [reply docs/api/webview-tag.md:611]: we can document it only taking `[number, number]` here, but `sendToFrame` is just forwarded to the webContents, which still takes the single value as well.

File: shell/renderer/api/electron_api_web_frame.cc (hunk 5 above)
Comment (miniak) [reply shell/renderer/api/electron_api_web_frame.cc]: done

File: docs/api/webview-tag.md (hunk 3 above)
Comment (nornagon) [reply docs/api/webview-tag.md]: Alrighty, let's keep it as is then.

**Review by nornagon — APPROVED** [review]

File: docs/api/webview-tag.md (hunk 4 above)
Comment (miniak) [reply docs/api/webview-tag.md:927]: yes, the reason why I kept the array is to allow `webview.sendToFrame(event.frameId, ...`, otherwise you would have to do `webview.sendToFrame([event.frameProcessId, event.frameRoutingId], ...`

**Comment by codebytere** [comment]
@miniak could you please rebase this? In addition - needs one more button-click API LGTM (@nornagon maybe?)

**Comment by miniak** [comment]
@codebytere conflicts resolved

**Comment by nornagon** [comment]
API LGTM. button click is a bit pointless b/c it gets reset by any rebase or other change.

**Review by jkleinsc — APPROVED** [review]

**Comment by miniak** [comment]
@codebytere can we merge this now?
